source/brailleDisplayDrivers/nattiqbraille.py

```diff
@@ -0,0 +1,143 @@
+import hwIo
+
+BAUD_RATE = 10000000
+INIT_TAG = b"0"
+RESET_TAG = b"reset"
+# Initialization response id
+INIT_RESP = 0
+# Keys response id
+ROUTE_RESP = 1
+UP_KEY_RESP = 2
+DOWN_KEY_RESP = 3
+RIGHT_KEY_RESP = 4
+LEFT_KEY_RESP = 5
+# Keys pressed id
+UP_KEY_PRESS = 1
+DOWN_KEY_PRESS = 2
+RIGHT_KEY_PRESS = 3
+LEFT_KEY_PRESS = 4
+
+
+class BrailleDisplayDriver(braille.BrailleDisplayDriver):
+	name = "nattiqbraille"
+	# Translators: Names of braille displays
+	description = _("Nattiq nBraille")
+	isThreadSafe = True
+
+	@classmethod
+	def getManualPorts(cls):
+		return braille.getSerialPorts()
+
+	def __init__(self, port="auto"):
+		super(BrailleDisplayDriver, self).__init__()
+		self._serial = None
+		for portType, portId, port, portInfo in self._getTryPorts(port):
+			log.debug("Checking port %s for a Nattiq nBraille", port)
+			try:
+				self._serial = hwIo.Serial(
+					port, baudrate=BAUD_RATE, timeout=self.timeout, writeTimeout=self.timeout,
+					parity=serial.PARITY_NONE, onReceive=self._onReceive
+				)
+			except EnvironmentError:
+				log.debugWarning("", exc_info=True)
+				continue
+			# Check for cell information
+			if self._describe():
+				log.debug("Nattiq nBraille found on %s with %d cells", port, self.numCells)
+				break
+			else:
+				self._serial.close()
+		else:
+			raise RuntimeError("Can't find a Nattiq nBraille device (port = %s)" % port)
+
+	def terminate(self):
+		try:
+			super(BrailleDisplayDriver, self).terminate()
+		finally:
+			self._serial.write(RESET_TAG)
+			self._serial.close()
+			self._serial = None
+
+	def _describe(self):
+		self.numCells = 0
+		log.debug("Writing reset tag")
+		self._serial.write(RESET_TAG)
+		self._serial.waitForRead(self.timeout * 10)
+		log.debug("Writing init tag")
+		self._serial.write(INIT_TAG)
+		self._serial.waitForRead(self.timeout * 10)
+		# If a valid response was received, _onReceive will have set numCells.
+		if self.numCells:
+			return True
+		log.debug("Not a Nattiq nBraille")
+		return False
+
+	def _onReceive(self, command):
+		if int(command) == INIT_RESP:
+			CELLS_NUM = self._serial.read(2)
+			self.numCells = int(CELLS_NUM)
+		elif int(command) == ROUTE_RESP:
+			ROUTE_KEY = self._serial.read(2)
+			inputCore.manager.executeGesture(RoutingInputGesture(int(ROUTE_KEY)))
+		elif int(command) == UP_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(UP_KEY_PRESS))
+			log.debug("Up Key Pressed")
+		elif int(command) == DOWN_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(DOWN_KEY_PRESS))
+			log.debug("Down Key Pressed")
+		elif int(command) == RIGHT_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(RIGHT_KEY_PRESS))
+			log.debug("Right Key Pressed")
+		elif int(command) == LEFT_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(LEFT_KEY_PRESS))
+			log.debug("Left Key Pressed")
+
+	def display(self, cells):
+		cells = "-".join(str(cell) for cell in cells)
+		log.debug(cells)
+		self._serial.write(cells.encode())

```

Comment:
In Python 3, there is a strict separation between strings and bytes objects. Strings assume you're working with textual data, which is not the case when writing cells to a braille display, so my personal  preference would be avoiding encoding like this.

If the dashes are really important, you can do:

`b"-".join(cell.to_bytes(1, "little", signed=False) for cell in cells)`

If the dashes can be ommited, you can just use `bytes(cells)` to convert the list of integers to bytes.

Reply:
It is still messing something up, because when I use the cells as byte and then send them, for some reasons the unit detects nothing at all, it just put the whole braille line as clear or empty.
While when using the cells as string data, then encoding them for the writing, the braille works fine on the unit. 
it might be something to do with how the unit interrupts the data itself.

Reply:
Ugh, it's because my assumptions where wrong. Of course, str(x) just returns a string with x, not the byte value of x. That also clarifies the dashes. Unfortunately, bytes(x) returns a bytes object length x, not a bytes object containing x.

I think I'd leave is as is, but may be @feerrenrut has additional ideas.

Reply:
![rsz_a_20200218_112722](https://user-images.githubusercontent.com/60925266/74713723-963b2880-5242-11ea-9bf6-044155c5aa2a.jpg)
So far, everything is working as it should with input and output.
It will be great to have more ideas about the optimal code, so that hopefully this driver works perfectly and does not create problems in the future
@leonardder  Thank you for the review.